### PR TITLE
python3Packages.emborg: init at 1.34

### PR DESCRIPTION
**Pull request by jpetrucciani**
###### Description of changes

I would like to be able to use [emborg](https://github.com/KenKundert/emborg) to control [borgbackup](https://www.borgbackup.org/) (which is already in nixpkgs!) from within nixpkgs! This also adds a handful of packages that are dependencies for this to be possible

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/emborg/default.nix [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  # this disables testing fuse mounts
+  MISSING_DEPENDENCIES = "fuse";
+  postPatch = ''
+    patchShebangs .
+  '';
+
+  # this patch fixes a whitespace issue in the message that a test is expecting
+  patches = [
+    (fetchpatch {
+      url = "https://github.com/jpetrucciani/emborg/commit/afac6d1ddcecdb4bddbec87b6c8eed4cfbf4ebf9.diff";
```
Comment (fabaff) [inline pkgs/development/python-modules/emborg/default.nix:57]: This should point to the upstream commit, not a commit in a fork. 

```suggestion
  patches = [
    (fetchpatch {
        # this patch fixes a whitespace issue in the message that a test is expecting, https://github.com/KenKundert/emborg/pull/67
      url = "https://github.com/KenKundert/emborg/commit/afac6d1ddcecdb4bddbec87b6c8eed4cfbf4ebf9.diff";
```

File: pkgs/development/python-modules/emborg/default.nix [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      sha256 = "3xg2z03FLKH4ckmiBZqE1FDjpgjgdO8OZL1ewrJlQ4o=";
+    })
+  ];
+
+  meta = with lib; {
+    description = "Interactive command line interface to Borg Backup";
+    homepage = "https://github.com/KenKundert/emborg";
+    license = licenses.gpl3;
```
Comment (fabaff) [inline pkgs/development/python-modules/emborg/default.nix:65]:
```suggestion
    license = licenses.gpl3Plus;
```

See https://github.com/KenKundert/emborg/blob/master/pyproject.toml#L15

File: pkgs/development/python-modules/parametrize-from-file/default.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    inherit version;
+    pname = "parametrize_from_file";
+    sha256 = "1c91j869n2vplvhawxc1sv8km8l53bhlxhhms43fyjsqvy351v5j";
+  };
+
+  format = "flit";
+  pythonImportsCheck = [ "parametrize_from_file" ];
+
+  checkInputs = [
+    coveralls
```
Comment (fabaff) [inline pkgs/development/python-modules/parametrize-from-file/default.nix:30]: Please patch coverage out this doesn't add value for us.

File: pkgs/development/python-modules/parametrize-from-file/default.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    more-itertools
+    nestedtext
+    pyyaml
+    tidyexc
+    toml
+  ];
+
+  meta = with lib; {
+    description = "Read unit test parameters from config files";
+    homepage = "https://github.com/kalekundert/parametrize_from_file";
```
Comment (fabaff) [inline pkgs/development/python-modules/parametrize-from-file/default.nix:46]:
```suggestion
    homepage = "https://github.com/kalekundert/parametrize_from_file";
    changelog = "https://github.com/kalekundert/parametrize_from_file/blob/v${version}/CHANGELOG.md";
```

File: pkgs/development/python-modules/quantiphy/default.nix [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+, parametrize-from-file
+, setuptools
+, voluptuous
+, quantiphy-eval
+, rkm-codes
+}:
+
+buildPythonPackage rec {
+  pname = "quantiphy";
+  version = "2.18";
```
Comment (fabaff) [inline pkgs/development/python-modules/quantiphy/default.nix:16]:
```suggestion
  version = "2.18";

   disabled = pythonOlder "3.6";
```

See https://github.com/KenKundert/quantiphy/blob/master/pyproject.toml#L30. I would suggest to stick to support Python releases for all new module. Thus, `disabled = pythonOlder "3.7";`.

File: pkgs/development/python-modules/quantiphy/default.nix [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    setuptools
+    voluptuous
+  ];
+  propagatedBuildInputs = [
+    quantiphy-eval
+    rkm-codes
+  ];
+
+  meta = with lib; {
+    description = "physical quantities (numbers with units)";
```
Comment (fabaff) [inline pkgs/development/python-modules/quantiphy/default.nix:42]:
```suggestion
    description = "Module for physical quantities (numbers with units)";
```

File: pkgs/development/python-modules/rkm-codes/default.nix [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+buildPythonPackage rec {
+  pname = "rkm-codes";
+  version = "0.5";
+
+  src = fetchFromGitHub {
+    owner = "KenKundert";
+    repo = "rkm_codes";
+    rev = "v${version}";
+    sha256 = "sha256-r4F72iHxH7BoPtgYm1RD6BeSZszKRrpeBQccmT4wzuw=";
+  };
```
Comment (fabaff) [inline pkgs/development/python-modules/rkm-codes/default.nix:16]:
```suggestion
  };

```

File: pkgs/development/python-modules/rkm-codes/default.nix [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  buildInputs = [ pytest-runner ];
```
Comment (fabaff) [inline pkgs/development/python-modules/rkm-codes/default.nix:17]: `pytest-runner` can in most cases be patched out.

File: pkgs/development/python-modules/rkm-codes/default.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+  # this has a circular dependency on quantiphy
+  preBuild = ''
+    sed -i '/quantiphy/d' ./setup.py
+  '';
+
+  # tests require quantiphy import
+  doCheck = false;
```
Comment (fabaff) [inline pkgs/development/python-modules/rkm-codes/default.nix:25]: Please add `pythonImportsCheck`.

File: pkgs/development/python-modules/rkm-codes/default.nix [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+  meta = with lib; {
+    description = "QuantiPhy support for RKM codes";
+    homepage = "https://nurdletech.com/linux-utilities/rkm_codes";
```
Comment (fabaff) [inline pkgs/development/python-modules/rkm-codes/default.nix:29]: The URL gives me a 404.

```suggestion
    homepage = "https://github.com/kenkundert/rkm_codes/";
```

File: pkgs/development/python-modules/rkm-codes/default.nix [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    license = licenses.gpl3;
```
Comment (fabaff) [inline pkgs/development/python-modules/rkm-codes/default.nix:30]:
```suggestion
    license = licenses.gpl3Plus;
```

See https://github.com/KenKundert/rkm_codes/blob/master/pyproject.toml#L14

File: pkgs/development/python-modules/shlib/default.nix [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  src = fetchFromGitHub {
+    owner = "KenKundert";
+    repo = "shlib";
+    rev = "v${version}";
+    sha256 = "sha256-2fwRxa64QXKJuhYwt9Z4BxhTeq1iwbd/IznfxPUjeSM=";
+  };
+
+  pythonImportsCheck = [ "shlib" ];
+  propagatedBuildInputs = [ braceexpand inform ];
+
```
Comment (fabaff) [inline pkgs/development/python-modules/shlib/default.nix:21]: Looks like that upstream is running the tests with `pytest`. Please check if we can run them with `pytestCheckHook`.

File: pkgs/development/python-modules/shlib/default.nix [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  meta = with lib; {
+    description = "shell library";
+    homepage = "https://nurdletech.com/linux-utilities/shlib";
```
Comment (fabaff) [inline pkgs/development/python-modules/shlib/default.nix:24]: The URL give me a 404.

```suggestion
    homepage = "https://github.com/KenKundert/shlib";
    changelog = "https://github.com/KenKundert/shlib/releases/tag/v${version}";
```

File: pkgs/development/python-modules/shlib/default.nix [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    license = licenses.gpl3;
```
Comment (fabaff) [inline pkgs/development/python-modules/shlib/default.nix:25]:
```suggestion
    license = licenses.gpl3Plus;
```

File: pkgs/development/python-modules/tidyexc/default.nix [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    inherit pname version;
+    sha256 = "1gl1jmihafawg7hvnn4xb20vd2x5qpvca0m1wr2gk0m2jj42yiq6";
+  };
+
+  format = "flit";
+  pythonImportsCheck = [ "tidyexc" ];
+
+  meta = with lib; {
+    description = "Raise rich, helpful exceptions";
+    homepage = "https://github.com/kalekundert/tidyexc";
```
Comment (fabaff) [inline pkgs/development/python-modules/tidyexc/default.nix:20]:
```suggestion
    homepage = "https://github.com/kalekundert/tidyexc";
    changelog = "https://github.com/kalekundert/tidyexc/blob/v${version}/CHANGELOG.md";
```

File: pkgs/development/python-modules/rkm-codes/default.nix (hunk 9 above)
Comment (jpetrucciani) [reply pkgs/development/python-modules/rkm-codes/default.nix:25]: This is one that was confusing to me - the `rkm-codes` and `quantiphy-eval` packages both are dependencies of `quantiphy` (at least for the tests), but they themselves have `quantiphy` as a dependency. Have you seen other packages like this? If so, how do you handle that within nixpkgs? Attempting to add an imports check here will cause an attempt to import `quantiphy`

File: pkgs/development/python-modules/parametrize-from-file/default.nix (hunk 3 above)
Comment (jpetrucciani) [reply pkgs/development/python-modules/parametrize-from-file/default.nix]: done!

**Comment by jpetrucciani** [comment]
@fabaff thanks for the thorough review! I've made all the requested changes except one - the `pythonImportsCheck` on `rkm-codes`. not sure how to best handle that one 🤔

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 207188` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>emborg (python310Packages.emborg)</li>
    <li>python310Packages.parametrize-from-file</li>
    <li>python310Packages.quantiphy</li>
    <li>python310Packages.quantiphy-eval</li>
    <li>python310Packages.rkm-codes</li>
    <li>python310Packages.shlib</li>
    <li>python310Packages.tidyexc</li>
    <li>python39Packages.emborg</li>
    <li>python39Packages.parametrize-from-file</li>
    <li>python39Packages.quantiphy</li>
    <li>python39Packages.quantiphy-eval</li>
    <li>python39Packages.rkm-codes</li>
    <li>python39Packages.shlib</li>
    <li>python39Packages.tidyexc</li>
  </ul>
</details>

File: pkgs/development/python-modules/rkm-codes/default.nix (hunk 9 above)
Comment (fabaff) [reply pkgs/development/python-modules/rkm-codes/default.nix:37]: Circular dependencies are always a pain. There were some workarounds present involving setting `doCheck = false;` in `python-modules.nix` and then inherit `doCheck`. Like for `sanic`:

```nix
  pytest-sanic = callPackage ../development/python-modules/pytest-sanic {
    sanic = self.sanic.override { doCheck = false; };
  };
```

Often there is not much value.

**Review by fabaff — APPROVED** [review]
Looks good to me.

**Pull request merged**